// @0xpolygonid/js-sdk/dist/esm/storage/shared/credential-storage.js
import { W3CCredential } from '../../verifiable';
import { StandardJSONCredentialsQueryFilter } from '../filters';
/**
 * Implementation of ICredentialStorage with KV Data source
 *
 * @export
 * @beta
 * @class CredentialStorage
 * @implements {ICredentialStorage}
 */
export class CredentialStorage {
    /**
     * Creates an instance of CredentialStorage.
     * @param {IDataSource<W3CCredential>} _dataSource - W3CCredential credential KV data source
     */
    constructor(_dataSource) {
        this._dataSource = _dataSource;
    }
    /** {@inheritdoc ICredentialStorage.listCredentials } */
    async listCredentials() {
        let creds = await this._dataSource.load();
        creds = creds.map((cred) => (cred ? Object.assign(new W3CCredential(), cred) : undefined));
        return creds;
    }
    /** @inheritdoc */
    async saveCredential(credential) {
        return this._dataSource.save(credential.id, credential);
    }
    /** {@inheritdoc ICredentialStorage.listCredentials } */
    async saveAllCredentials(credentials) {
        for (const credential of credentials) {
            await this._dataSource.save(credential.id, credential);
        }
    }
    /** {@inheritdoc ICredentialStorage.listCredentials } */
    async removeCredential(id) {
        return this._dataSource.delete(id);
    }
    /** {@inheritdoc ICredentialStorage.listCredentials } */
    async findCredentialById(id) {
        const cred = await this._dataSource.get(id);
        return cred ? Object.assign(new W3CCredential(), cred) : undefined;
    }
    /** {@inheritdoc ICredentialStorage.listCredentials }
     * uses JSON query
     */
    async findCredentialsByQuery(query) {
        const filters = StandardJSONCredentialsQueryFilter(query);
        let creds = (await this._dataSource.load()).filter((credential) => filters.every((filter) => filter.execute(credential)));
        creds = creds.map((cred) => (cred ? Object.assign(new W3CCredential(), cred) : undefined));
        return creds;
    }
}
/**
 * key for storage
 *
 * @static
 */
CredentialStorage.storageKey = 'credentials';
